src/helpers/pivot/spreadsheet_pivot/data_entry_spreadsheet_pivot.ts

```diff
@@ -0,0 +1,254 @@
+  const colName = columns[index].nameWithGranularity;
+  const groups = groupBy(dataEntries, column);
+  const orderedKeys = orderDataEntriesKeys(groups, columns[index]);
+  const columnsTree: ColumnsTree = [];
+  for (const value of orderedKeys) {
+    columnsTree.push({
+      value,
+      field: colName,
+      children: dateEntriesToColumnsTree(groups[value] || [], columns, index + 1),
+      width: 0,
+    });
+  }
+  return columnsTree;
+}
+/**
+ * Compute the width of each node in the column tree.
+ * The width of a node is the sum of the width of its children.
+ * For leaf nodes, the width is the number of measures.
+ */
+function computeWidthOfColumnsNodes(tree: ColumnsTree, measureCount: number) {
+  for (const key in tree) {
+    const node = tree[key];
+    if (Object.keys(node.children).length === 0) {
+      node.width = measureCount;
+    } else {
+      computeWidthOfColumnsNodes(node.children, measureCount);
+      node.width = Object.values(node.children).reduce((acc, child) => acc + child.width, 0);
+    }
+  }
+}
+
+/**
+ * Convert the columns tree to the columns
+ */
+function columnsTreeToColumns(
+  mainTree: ColumnsTree,
+  definition: SpreadsheetPivotRuntimeDefinition
+): SPTableColumn[][] {
+  const columnNames = definition.columns.map((col) => col.nameWithGranularity);
+  const height = columnNames.length;
+  const measureCount = definition.measures.length;
+
+  const headers: SPTableColumn[][] = new Array(height).fill(0).map(() => []);
+
+  function generateTreeHeaders(tree: ColumnsTree, rowIndex: number, val: string[]) {
+    const row = headers[rowIndex];
+    for (const node of Object.values(tree)) {
+      const localVal = val.concat([node.value]);
+      const cell: SPTableColumn = {
+        fields: columnNames.slice(0, rowIndex + 1),
+        values: localVal,
+        width: node.width,
+        offset: 0,
+      };
+      row.push(cell);
+      if (rowIndex <= height - 1) {
+        generateTreeHeaders(node.children, rowIndex + 1, localVal);
+      }
+    }
+  }
+  generateTreeHeaders(mainTree, 0, []);
+  const hasColGroupBys = columnNames.length > 0;
+
+  // 2) generate measures row
+  const measureRow: SPTableColumn[] = [];
+
+  if (hasColGroupBys) {
+    headers[headers.length - 1].forEach((cell) => {
+      definition.measures.forEach((measure) => {
+        const measureCell = {
+          fields: [...cell.fields, "measure"],
+          values: [...cell.values, measure.name],
+          width: 1,
+          offset: 0,
+        };
+        measureRow.push(measureCell);
+      });
+    });
+  }
+  // Add the totals of the measures
+  definition.measures.forEach((measure) => {
+    const measureCell = {
+      fields: ["measure"],
+      values: [measure.name],
+      width: 1,
+      offset: 0,
+    };
+    measureRow.push(measureCell);
+  });
+  headers.push(measureRow);
+  // 3) Add the total cell
+  if (headers.length === 1) {
+    headers.unshift([]); // Will add the total there
+  }
```

Comment:
Cannot do it easily, because `generateTreeHeaders` use the index of the row in the headers, so this function will have to take into account if we already add an empty header or not.
Unshift is O(n), but n in this case is the number of colGroupBys. It should stay at a reasonable number.
What do you think ?

Reply:
we can keep it, it's used only once so it should be fine